wgsl/extract-grammar.py

```diff
@@ -510,7 +761,321 @@ def read_spec(options):
+                    scan_in["content"][0:-2],
+                    syntax_dict[scan_in["value"]],
+                )["members"][-1]["type"] = "choice"
+                s_i += 1
+            elif s_v == "(":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-1],
+                    syntax_dict[scan_in["value"]],
+                )["members"].append({"type": "sequence", "members": []})
+                scan_in["content"].append(0)
+                s_i += 1
+            elif s_v == ")":
+                scan_in["content"] = scan_in["content"][0:-1]
+                s_i += 1
+            else:
+                s_i += 1
+        else:
+            s_i += 1
+        if s_i < len(bnfdialect_data):
+            s_v = bnfdialect_data[s_i]
+        else:
+            s_v = ""
+
+    def reproduce_rule(production):
+        if production["type"] in ["sequence", "choice", ""]:
+            if len(production["members"]) == 1:
+                if production["members"][0]["type"] in ["token", "symbol", "pattern"]:
+                    return production["members"][0]
+                else:
+                    return reproduce_rule(production["members"][0])
+            else:
+                return {
+                    "type": production["type"],
+                    "members": [
+                        reproduce_rule(member) for member in production["members"]
+                    ],
+                }
+        elif production["type"].startswith("match_"):
+            if len(production["members"]) == 1:
+                if production["members"][0]["type"] == "sequence":
+                    return {
+                        "type": production["type"],
+                        "members": [
+                            reproduce_rule(member)
+                            for member in production["members"][0]["members"]
+                        ],
+                    }
+                else:
+                    return {
+                        "type": production["type"],
+                        "members": [
+                            reproduce_rule(member) for member in production["members"]
+                        ],
+                    }
+            else:
+                return {
+                    "type": production["type"],
+                    "members": [
+                        reproduce_rule(member) for member in production["members"]
+                    ],
+                }
+        else:
+            return production
+
+    # Simplify the syntax dictionary
+    for rule_name in syntax_dict.keys():
+        syntax_dict[rule_name] = reproduce_rule(syntax_dict[rule_name])
+
+    if reproduce_bnfdialect_source(syntax_dict).strip() != bnfdialect_data.strip():
+        print("ERROR: Syntax source should match reproduction for styling and language")
+        sys.exit(1)
+
+    # Extract syntax to bikeshed fragments
```

Comment:
Put this in a different function, and invoke it separately.
The new function takes syntax_dict as an argument, and the Options object, so it knows what directory/files to update.